### ISSUE TITLE
Select all matches is #doingitwrong

### ISSUE THREAD

**Issue by pjv** [zed-industries/zed#21881]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

In the image below I've entered some ones separated by spaces and then one little stretch of three ones with no spaces between them.

If I select a single one and then execute `editor:select all matches` the selections and multiple cursors can be seen in this screenshot.

The error is that the string of three ones that are not separated by spaces are selected together as a single selection of `111` but that string is not a match to the original selection of `1`. There should be three separate selections there next to each other instead of one selection of `111`.

<img width="290" alt="SCR-20241211-nkal-2" src="https://github.com/user-attachments/assets/1616da48-36f9-47b2-b0c1-cfea43bfca48" />

### Environment

Zed: v0.166.0 (Zed Preview)
OS: macOS 14.7.1
Memory: 64 GiB
Architecture: x86_64

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary>

```

```
</details>

**Comment by failable** [comment]
@notpeter Isn't there only one cursor in line 5?

**Comment by notpeter** [comment]
Yes. Maybe there are alternate steps which trigger the original behavior, but I couldn't reproduce.

<img width="305" alt="Image" src="https://github.com/user-attachments/assets/e65092b7-90cc-41d9-979b-3edede2b3210" />

**Comment by failable** [comment]
> Yes. One cursor.

So I think this is still the issue? I think there should be lots of the cursors in line 5 based on the description of OP.

> There should be three separate selections there next to each other instead of one selection of 111.

**Comment by pjv** [comment]
@notpeter three things:

1. your demonstration still shows the issue. You should not have a single selection on line 5; you should have as many separate selections as you have 1's on that line.

2. in my original test above, i have 1 separated by spaces and then a short string of three 1's with no spaces, and then a couple more 1's with spaces. I believe that if a single 1 is selected and then `select all matches` is performed, there should be as many separate selections as there are individual 1 characters. The fact that my string of 3 ones in a row becomes a single selection (and your longer string of many 1's in a row becomes a single selection) is what i think is incorrect.

3. with Zed Preview 0.173.5, i now seem to be unable to make `select all matches` work at all. I select a character and execute `select all matches` and nothing happens at all.

**Comment by pjv** [comment]
here's a test file. When i do `select all matches` after selecting the first 1, i get a total of 5 selections. I believe there should be 7 selections.

[test-selections.txt](https://github.com/user-attachments/files/18769403/test-selections.txt)

**Comment by notpeter** [comment]
Ah. I see. I misread and now understand.
Yes. Zed currently merges adjacent selections.
I believe this is because overlapping selections lead to a bunch of problems.
But maybe we could preserve adjacent selections to match the behavior of vscode.
Opened that as a clean issue here:
- https://github.com/zed-industries/zed/issues/24748

Let me know if I'm still misunderstanding.

**Comment by pjv** [comment]
Yup, you got it.

I hope it's possible to preserve adjacent selections because if you wanted to change all those 1's to 2's by doing `editor:select all matches` and typing '2', you'd be SOL.
